### ISSUE TITLE
-res-type does not produce same output between bin and prob

### ISSUE THREAD

**Issue by jcohenadad** [neuropoly/spinalcordtoolbox#1329]
### Description

-res-type bin seems to produce larger segmentations than expected

### Steps to Reproduce

1. data: **20170516_issue1329**
2. command: ```sct_segment_graymatter -i x2D_MERGE.nii -s t2s_seg_manual.nii.gz -res-type bin```

**Expected behavior:** I would expect the same result between "bin" and "prob thresholded at 0.5"

**Actual behavior:** the output of bin seems a bit larger than "prob thresholded at 0.5"

### Versions

master/4d453e11c5f10e11a8ef8acc1b2979bb285dc5d7

### Additional Information

![webp net-gifmaker](https://cloud.githubusercontent.com/assets/2482071/26110950/c26bfda8-3a21-11e7-982a-3c1b89048967.gif)

~~~
sct_segment_graymatter -i x2D_MERGE.nii -s t2s_seg_manual.nii.gz -res-type bin

--
Spinal Cord Toolbox (master/4d453e11c5f10e11a8ef8acc1b2979bb285dc5d7)
Running /Users/julien/code/sct/scripts/sct_segment_graymatter.py -i x2D_MERGE.nii -s t2s_seg_manual.nii.gz -res-type prob

Create temporary folder...

Loading model...
  OK: intensities.pklz
  OK: fitted_data.pklz
  OK: fitted_model.pklz
  OK: slices.pklz
  844 slices in the model dataset
  model: pca
  21 components kept on 844

Pre-process data...
  Reorient...
  Denoise...
  OK: t2s_seg_manual_RPI.nii.gz

Create temporary folder...

Create temporary folder...

Check orientation...
.. RPI
sct_convert -i x2D_MERGE_RPI.nii -o tmp.170516101550_827819/data.nii
sct_convert -i t2s_seg_manual_RPI.nii.gz -o tmp.170516101550_827819/centerline.nii.gz

Reorient to RPI...

Get dimensions of data...
  512 x 512 x 16 x 1

Create mask...
sct_image -i mask_RPI.nii.gz -o mask.nii.gz -setorient RPI
WARNING: File mask.nii.gz already exists. Deleting it.

Generate output files...
  File created: mask_pre_crop.nii.gz

Remove temporary files...
rm -rf tmp.170516101550_827819/

Done! To view results, type:
fslview x2D_MERGE_RPI.nii mask_pre_crop.nii.gz -l Red -t 0.5 &

isct_crop_image -i x2D_MERGE_RPI.nii -o x2D_MERGE_RPI_pre_crop.nii -start 215,173,0 -end 290,295,15 -dim 0,1,2
Cropping the following region:
Dimension 0: 215 290
Dimension 1: 173 295
Dimension 2: 0 15

Done! To view results, type:
fslview x2D_MERGE_RPI_pre_crop.nii &

isct_crop_image -i t2s_seg_manual_RPI.nii.gz -o t2s_seg_manual_RPI_pre_crop.nii.gz -start 215,173,0 -end 290,295,15 -dim 0,1,2
Cropping the following region:
Dimension 0: 215 290
Dimension 1: 173 295
Dimension 2: 0 15

Done! To view results, type:
fslview t2s_seg_manual_RPI_pre_crop.nii.gz &

  Interpolate data to the model space...
  Mask data using the spinal cord segmentation...
WARNING: File im_ref.nii.gz already exists. Deleting it.
  Split along rostro-caudal direction...
  Load vertebral levels...
/Users/julien/code/sct/python/lib/python2.7/site-packages/numpy/core/fromnumeric.py:2889: RuntimeWarning: Mean of empty slice.
  out=out, **kwargs)
WARNING: cannot convert float NaN to integer
No level label found. Level will be set to 0 for this slice

Register target image to model data...
WARNING: File im_to_orient.nii.gz already exists. Deleting it.

Create temporary folder...

Input parameters:
  Source .............. src.nii.gz
  Destination ......... dest.nii.gz
  Init transfo ........ 
  Mask ................ 
  Output name ......... 
  Remove temp files ... 1
  Verbose ............. 1

Check if input data are 3D...

Create temporary folder...

Copying input data to tmp folder and convert to nii...
sct_convert -i src.nii.gz -o tmp.170516101559_240634/src.nii
sct_convert -i dest.nii.gz -o tmp.170516101559_240634/dest.nii
sct_convert -i src_seg.nii.gz -o tmp.170516101559_240634/src_seg.nii
sct_convert -i dest_seg.nii.gz -o tmp.170516101559_240634/dest_seg.nii
sct_image -i dest.nii -setorient RPI -o dest_RPI.nii
sct_image -i dest_seg.nii -setorient RPI -o dest_seg_RPI.nii

--
ESTIMATE TRANSFORMATION FOR STEP #0
Registration parameters:
  type ........... im
  algo ........... syn
  slicewise ...... 0
  metric ......... MI
  iter ........... 0
  smooth ......... 0
  laplacian ...... 0
  shrink ......... 1
  gradStep ....... 0.5
  init ........... 
  poly ........... 5
  dof ............ Tx_Ty_Tz_Rx_Ry_Rz
  smoothWarpXY ... 2

Estimate transformation
isct_antsRegistration --dimensionality 3 --transform syn[0.5,3,0] --metric MI[dest_RPI.nii,src.nii,1,32] --convergence 0 --shrink-factors 1 --smoothing-sigmas 0mm --restrict-deformation 1x1x0 --output [step0,src_regStep0.nii] --interpolation BSpline[3]  --verbose 1

--
ESTIMATE TRANSFORMATION FOR STEP #1

Apply transformation from previous step
sct_apply_transfo -i src_seg.nii -d dest_seg_RPI.nii -w warp_forward_0.nii.gz -o src_seg_reg.nii -x nn
Registration parameters:
  type ........... seg
  algo ........... centermassrot
  slicewise ...... 0
  metric ......... MeanSquares
  iter ........... 1
  smooth ......... 2
  laplacian ...... 0
  shrink ......... 1
  gradStep ....... 0.5
  init ........... 
  poly ........... 0
  dof ............ Tx_Ty_Tz_Rx_Ry_Rz
  smoothWarpXY ... 2

Smooth data
sct_maths -i src_seg_reg.nii -smooth 2,2,0 -o src_seg_reg_smooth.nii
sct_maths -i dest_seg_RPI.nii -smooth 2,2,0 -o dest_seg_RPI_smooth.nii

Create temporary folder...

Copy input data to temp folder...
sct_convert -i src_seg_reg_smooth.nii -o tmp.170516101602_648638/src.nii
sct_convert -i dest_seg_RPI_smooth.nii -o tmp.170516101602_648638/dest.nii

Get image dimensions of destination image...
  matrix size: 75 x 75 x 16
  voxel size:  0.3mm x 0.3mm x 1.0mm

Split input volume...

Split destination volume...
0/16.. 1/16.. 2/16.. 3/16.. 4/16.. 5/16.. 6/16.. 7/16.. 8/16.. 9/16.. 10/16.. 11/16.. 12/16.. 13/16.. 14/16.. 15/16.. 
Generate warping field...
 --> step1Warp.nii.gz

Generate warping field...
 --> step1InverseWarp.nii.gz

Move warping fields to parent folder...
mv step1Warp.nii.gz ../
mv step1InverseWarp.nii.gz ../

--
ESTIMATE TRANSFORMATION FOR STEP #2

Apply transformation from previous step
sct_apply_transfo -i src_seg.nii -d dest_seg_RPI.nii -w warp_forward_0.nii.gz,warp_forward_1.nii.gz -o src_seg_reg.nii -x nn
Registration parameters:
  type ........... seg
  algo ........... columnwise
  slicewise ...... 0
  metric ......... MeanSquares
  iter ........... 1
  smooth ......... 1
  laplacian ...... 0
  shrink ......... 1
  gradStep ....... 0.5
  init ........... 
  poly ........... 5
  dof ............ Tx_Ty_Tz_Rx_Ry_Rz
  smoothWarpXY ... 2

Smooth data
sct_maths -i src_seg_reg.nii -smooth 1,1,0 -o src_seg_reg_smooth.nii
sct_maths -i dest_seg_RPI.nii -smooth 1,1,0 -o dest_seg_RPI_smooth.nii

Create temporary folder...

Copy input data to temp folder...
sct_convert -i src_seg_reg_smooth.nii -o tmp.170516101604_867389/src.nii
sct_convert -i dest_seg_RPI_smooth.nii -o tmp.170516101604_867389/dest.nii

Get image dimensions of destination image...
  matrix size: 75 x 75 x 16
  voxel size:  0.3mm x 0.3mm x 1.0mm

Split input volume...

Split destination volume...

Estimate columnwise transformation...
0/16.. 1/16.. 2/16.. 3/16.. 4/16.. 5/16.. 6/16.. 7/16.. 8/16.. 9/16.. 10/16.. 11/16.. 12/16.. 13/16.. 14/16.. 15/16.. 
Generate warping field...
 --> step2Warp.nii.gz

Generate warping field...
 --> step2InverseWarp.nii.gz

Move warping fields to parent folder...
mv step2Warp.nii.gz ../
mv step2InverseWarp.nii.gz ../

Concatenate transformations...
sct_concat_transfo -w warp_forward_0.nii.gz,warp_forward_1.nii.gz,warp_forward_2.nii.gz -d dest.nii -o warp_src2dest.nii.gz
sct_concat_transfo -w warp_inverse_2.nii.gz,warp_inverse_1.nii.gz,warp_inverse_0.nii.gz -d dest.nii -o warp_dest2src.nii.gz

Apply transfo source --> dest...
sct_apply_transfo -i src.nii -o src_reg.nii -d dest.nii -w warp_src2dest.nii.gz -x linear

Apply transfo dest --> source...
sct_apply_transfo -i dest.nii -o dest_reg.nii -d src.nii -w warp_dest2src.nii.gz -x linear

Generate output files...
sct_convert -i tmp.170516101559_240634/src_reg.nii -o src_reg.nii.gz
  File created: src_reg.nii.gz
  File created: warp_src2dest.nii.gz
sct_convert -i tmp.170516101559_240634/dest_reg.nii -o dest_reg.nii.gz
  File created: dest_reg.nii.gz
  File created: warp_dest2src.nii.gz

Remove temporary files...
rm -rf tmp.170516101559_240634/

Finished! Elapsed time: 8s

To view results, type:
fslview src.nii.gz dest_reg.nii.gz &
fslview dest.nii.gz src_reg.nii.gz &


Normalize intensity of target image...

Project target image into the model reduced space...

Compute similarities between target slices and model slices using model reduced space...

Label fusion of model slices most similar to target slices...

Warp back segmentation into image space...
WARNING: File im_to_orient.nii.gz already exists. Deleting it.
WARNING: File im_to_orient.nii.gz already exists. Deleting it.

Create temporary folder...

Parse list of warping fields...
  Transfo #0: warp_dic2target.nii.gz

Get dimensions of data...
  16 x 75 x 75 x 1

Apply transformation...
isct_antsApplyTransforms -d 3 -i src.nii.gz -o src_reg.nii.gz -t warp_dic2target.nii.gz -r dest.nii.gz -n Linear

Done! To view results, type:
fslview dest.nii.gz src_reg.nii.gz &


Post-processing...
  Interpolate result back into original space...
  Reorient resulting segmentations to native orientation...

Save resulting GM and WM segmentations...
WARNING: File ./x2D_MERGE_gmseg.nii already exists. Deleting it.
WARNING: File ./x2D_MERGE_wmseg.nii already exists. Deleting it.

Finished! Elapsed time: 41s

Save quality control images...
QC output image: ./x2D_MERGE_axial_plane.png & ./x2D_MERGE_axial_plane_seg.png

Done! To view results, type:
fslview x2D_MERGE.nii ./x2D_MERGE_gmseg.nii -b 0.4,1 -l Red-Yellow -t 0.7 ./x2D_MERGE_wmseg.nii -b 0.4,1 -l Blue-Lightblue -t 0.7  & 
~~~

**Comment by SaraDupont** [comment]
The binarization is done right after the label fusion (i.e. still in the model space). The differences between the output of `-res-type bin` and `-res-type prob` thresholded at 0.5 come from the warping from the model space back to the original image space. I can move the thresholding after the warping to have the exact same result between both.

**Comment by jcohenadad** [comment]
yes, I would do the binarization after warping, to minimize interpolation errors

**Comment by SaraDupont** [comment]
Fixed in PR #1347
